include/boost/geometry/policies/predicate_based_interrupt_policy.hpp

```diff
@@ -43,12 +43,12 @@ struct stateless_predicate_based_interrupt_policy
     template <typename Range>
     inline bool apply(Range const& range)
     {
-        // if there is at least one unacceptable turn in the range, return false
-        has_intersections = !detail::check_iterator_range
-            <
-                IsAcceptableTurnPredicate,
-                AllowEmptyTurnRange
-            >::apply(boost::begin(range), boost::end(range));
+        has_intersections = ! (boost::end(range) == std::find_if(
+            boost::begin(range), 
+            boost::end(range),
+            []( auto const& turn ){ 
+                return ! IsAcceptableTurnPredicate::apply(turn); }
+        ) || (AllowEmptyTurnRange && boost::empty(range)));
```

Comment:
This is quite complex. How about:
```
auto const end = boost::end(range);
bool const has_unacceptable_turn
    = std::find_if(boost::begin(range), end,
                   []( auto const& turn ) { 
                       return ! IsAcceptableTurnPredicate::apply(turn);
                   }) != end;
has_intersections = ! has_unacceptable_turn
                 || (AllowEmptyTurnRange && boost::empty(range)));
```

or something more clear like that.

Also the comparison operator wasn't correct, right? With `==` true would be returned if all turns were acceptable or for empty range. So `has_intersection` would be true for unacceptable turns and non-empty or `AllowEmptyTurnRange` and empty. Or am I missing something? Does all of the tests pass?

Reply:
Ok, I know what I missed, parenthesis.

Reply:
There was one negation missing, so maybe it's better top keep these parenthesis and negate the whole expression as you did. For that `has_unacceptable_turn` would have to be renamed to something like `has_acceptable_turns_or_empty`.

Reply:
I think there is one more wrong negation still in the your proposed code. Indeed it is complicated. The correct should be
```
bool const has_unacceptable_turn
    = std::find_if(boost::begin(range), end,
                   []( auto const& turn ) { 
                       return ! IsAcceptableTurnPredicate::apply(turn);
                   }) != end;
has_intersections = has_unacceptable_turn
    && !(AllowEmptyTurnRange && boost::empty(range));
```
with only one negation as in the original but less complicated as you propose.

Reply:
Well, I don't know, the original comment states:
```
// if there is at least one unacceptable turn in the range, return false
```
So AFAIU if `has_unacceptable_turn` is `true` then `has_intersections` should be `false`, right? So either something is wrong with the naming or with `find_if`. Or I'm missing something.

Reply:
See the original code too, it seems that the original comment is wrong it should state 
```
// if there is at least one unacceptable turn in the range, return false
```

Reply:
There are two booleans here:
- `has_intersections` which is set to `true` if there are unacceptable turns or if the ring is empty when it shouldn't, this is used in `in_valid` and may have a different meaning than in other parts of the library, here it means that there the turns that were found are invalid [1].
- return value which indicates if processing of turns should end [2], AFAIU `true` means stop [3]

So if there are unacceptable turns found (or empty ring when it shouldn't) `has_intersections` should be `true` and this should also be the return value. So you're correct and I was missing that here `has_intersections` is used to return error condition.

[1] https://github.com/boostorg/geometry/blob/develop/include/boost/geometry/algorithms/detail/is_valid/has_valid_self_turns.hpp#L92-L95
[2] https://github.com/boostorg/geometry/blob/develop/include/boost/geometry/algorithms/detail/is_valid/has_valid_self_turns.hpp#L80-L83
[3] https://github.com/boostorg/geometry/blob/develop/include/boost/geometry/algorithms/detail/overlay/self_turn_points.hpp#L48-L59